R/settings.R

```diff
@@ -77,6 +78,73 @@ read_config_file <- function(config_file) {
   config
 }
 
+validate_config_file <- function(config, config_file, defaults) {
+  matched <- names(config) %in% names(defaults)
+  if (!all(matched)) {
+    warning("Found unused settings in config '", config_file, "': ", toString(names(config)[!matched]))
+  }
+
+  # exclude can flexibly accept list() or character()
+  check_types <- setdiff(intersect(names(config), names(defaults)), "exclusions")
+  for (setting in check_types) {
```

Comment:
Could also check the lengths for string-type settings.

Or does this work?
`exclude: c("#nolint", "#mynolint")`

Reply:
No, it's passed to `rex::re_matches()`:

https://github.com/r-lib/lintr/blob/b49ee9dac3eddb9285648c9e2b5f15e85659ee3a/R/exclude.R#L140

Which AFAICT only accepts scalar. I suppose we could accept that (just collapse with `|`) but that seems like asking for trouble, user can do that themselves if needed.

Reply:
So we should validate for length 1.

Reply:
Refactored the approach, I think it balances complexity with extensibility pretty well. Basically we class the types of options there are and run the validator on each of those. Now there are 5 types of options: (1) character string, (2) regex (=character string+), (3) TRUE/FALSE, (4) linters (5) exclusions.